### PR TITLE
Added Beautiful Subgrids

### LINKED ISSUE #209: [Feature Request] Beautiful Subgrids - Recently Asked Interview Question at Apple
### Is your feature request related to a problem? Please describe.
This problem was recently asked in a coding interview at Apple, emphasizing the need for optimized solutions that can handle large inputs and detect complex patterns in a 2D grid. The problem is related to efficiently finding "Beautiful Subgrids" in a grid matrix. A subgrid is considered "beautiful" if it has a specific property, such as symmetry or all equal values in the subgrid. The challenge is to detect and count such subgrids within larger matrices efficiently, which can be computationally expensive as the grid size increases.

### Describe the solution you'd like
The solution should include an algorithm that can:

- Identify all subgrids of various sizes within a given grid.
- Efficiently check if each subgrid meets the "beautiful" criteria. This could involve checking for symmetry, uniform values, or specific patterns.
- Optimize the approach using techniques such as prefix sums, dynamic programming, or matrix-based algorithms to reduce time complexity. The goal is to ensure the solution scales well for large grids.

One approach might involve:

- Preprocessing the grid to store properties like row and column parity or prefix sums.
- Matrix DP to calculate properties of subgrids as we expand the window over the matrix, avoiding recomputation for overlapping regions.
- Efficient pattern matching to detect symmetry or other grid properties.

Would solve the problem in C++, Java, Python, Javascript, Go and clearly explain each and every step and add key points to mention at the interview as well!

---

**Would you like to work on this feature?**  
- [x] Yes  
- [ ] No

### PR DESCRIPTION
**Pull request by J-B-Mugundh**
### Description
Added the Beautiful Subgrids problem which involves finding subgrids within a larger 2D grid that possess certain "beautiful" properties, such as symmetry or uniformity in values. The challenge is to efficiently count such subgrids, especially as the grid size increases.

Fixes #209 

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by ajay-dhangar** [comment]
@J-B-Mugundh, please sync your fork with the latest changes from the main repository. I've resolved the `backtracking-questions.md` error at line 56.

Here's what you need to do:
1. Fetch the latest updates from the main repository:
   ```
   git fetch upstream
   git merge upstream/main
   ```
2. Push the merged changes to your fork:
   ```
   git push origin your-branch
   ```
3. After syncing, resolve any remaining issues.

Please don't hesitate to let me know if you encounter any problems.

**Comment by J-B-Mugundh** [comment]
@ajay-dhangar Updated bro! Also, this was labelled as level 2 in the issue. Could you please change it to level 2 here?

File: docs/dynamic-programming/Beautiful-Subgrids.md [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+tags: [dsa, dp, grid problems, matrix]
+---
+
+## Problem Definition
+
+The Beautiful Subgrids problem involves finding subgrids within a larger 2D grid that possess certain "beautiful" properties, such as symmetry or uniformity in values. The challenge is to efficiently count such subgrids, especially as the grid size increases.
+
+### Constraints
+
+- The dimensions of the grid can be large, potentially reaching up to \(10^5\) rows and \(10^5\) columns.
```
Comment (ajay-dhangar) [inline docs/dynamic-programming/Beautiful-Subgrids.md]: replace `v` with `$10^5$`

File: docs/dynamic-programming/Beautiful-Subgrids.md [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+  
+2. **Check for Beauty**: For each subgrid, check if it satisfies the beautiful condition (e.g., all elements being equal).
+
+3. **Count Valid Subgrids**: Maintain a counter to count how many of these subgrids are considered beautiful.
+
+## Complexity Analysis
+
+- **Time Complexity**: The worst-case time complexity is \(O(m^2 . n^2)\), where \(m\) is the number of rows and \(n\) is the number of columns in the grid.
+- **Space Complexity**: The space complexity is \(O(1)\) for auxiliary storage, but this could increase if storing subgrids is necessary.
+
```
Comment (ajay-dhangar) [inline docs/dynamic-programming/Beautiful-Subgrids.md]: - replace `\(O(m^2 . n^2)\)` with `$O(m^2 . n^2)$`,
- `\(m\)` $\rightarrow$ `$m$`
- `\(n\)` $\rightarrow$ `$n$`
- `\(O(1)\)` $\rightarrow$ `$O(1)$`

**Review by ajay-dhangar — CHANGES_REQUESTED** [review]
update your code based on suggestions

**Comment by ajay-dhangar** [comment]
> @ajay-dhangar Updated bro! Also, this was labelled as level 2 in the issue. Could you please change it to level 2 here?

I assign you a level based on the quality of your content, not the quantity. This is because we need quality content, not just a lot of content.

**Comment by J-B-Mugundh** [comment]
> > @ajay-dhangar Updated bro! Also, this was labelled as level 2 in the issue. Could you please change it to level 2 here?
> 
> I assign you a level based on the quality of your content, not the quantity. This is because we need quality content, not just a lot of content.

Completely agree your point and I believe the same. Thank you so much for your feedback. Also, I would appreciate if you could guide me to make better contents in future.

**Comment by J-B-Mugundh** [comment]
> > @ajay-dhangar Updated bro! Also, this was labelled as level 2 in the issue. Could you please change it to level 2 here?
> 
> I assign you a level based on the quality of your content, not the quantity. This is because we need quality content, not just a lot of content.

Okay bro! I also wish the same. Thanks for helping me

**Review by ajay-dhangar — APPROVED** [review]
Done

**Pull request merged**